# Machine Learning Engineer Nanodegree
## Supervised Learning
## Project 2: Building a Student Intervention System

Welcome to the second project of the Machine Learning Engineer Nanodegree! In this notebook, some template code has already been provided for you, and it will be your job to implement the additional functionality necessary to successfully complete this project. Sections that begin with **'Implementation'** in the header indicate that the following block of code will require additional functionality which you must provide. Instructions will be provided for each section and the specifics of the implementation are marked in the code block with a `'TODO'` statement. Please be sure to read the instructions carefully!

In addition to implementing code, there will be questions that you must answer which relate to the project and your implementation. Each section where you will answer a question is preceded by a **'Question X'** header. Carefully read each question and provide thorough answers in the following text boxes that begin with **'Answer:'**. Your project submission will be evaluated based on your answers to each of the questions and the implementation you provide.  

>**Note:** Code and Markdown cells can be executed using the **Shift + Enter** keyboard shortcut. In addition, Markdown cells can be edited by typically double-clicking the cell to enter edit mode.

### Question 1 - Classification vs. Regression
*Your goal for this project is to identify students who might need early intervention before they fail to graduate. Which type of supervised learning problem is this, classification or regression? Why?*

**Answer: **

This is a **classification** task, because the goal is to sort elements (students) into classes (likely to graduate; not likely to graduate). The task isn't to report the likelihood that a given student will fail, which would be a regression task.

## Exploring the Data
Run the code cell below to load necessary Python libraries and load the student data. Note that the last column from this dataset, `'passed'`, will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.

In [1]:
# Import libraries
import numpy as np
import pandas as pd
from time import time
from sklearn.metrics import f1_score

# Read student data
student_data = pd.read_csv("student-data.csv")
print "Student data read successfully!"

Student data read successfully!


### Implementation: Data Exploration
Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
- The total number of students, `n_students`.
- The total number of features for each student, `n_features`.
- The number of those students who passed, `n_passed`.
- The number of those students who failed, `n_failed`.
- The graduation rate of the class, `grad_rate`, in percent (%).


In [4]:
# TODO: Calculate number of students
n_students = len(student_data)

# TODO: Calculate number of features
n_features = len(student_data.columns) - 1 # Excludes the 'passed' column

# TODO: Calculate passing students
n_passed = len(student_data[student_data['passed'] == 'yes'])

# TODO: Calculate failing students
n_failed = len(student_data[student_data['passed'] == 'no'])

# TODO: Calculate graduation rate
grad_rate = float(n_passed) / n_students * 100

# Print the results
print "Total number of students: {}".format(n_students)
print "Number of features: {}".format(n_features)
print "Number of students who passed: {}".format(n_passed)
print "Number of students who failed: {}".format(n_failed)
print "Graduation rate of the class: {:.2f}%".format(grad_rate)

Total number of students: 395
Number of features: 30
Number of students who passed: 265
Number of students who failed: 130
Graduation rate of the class: 67.09%


## Preparing the Data
In this section, we will prepare the data for modeling, training and testing.

### Identify feature and target columns
It is often the case that the data you obtain contains non-numeric features. This can be a problem, as most machine learning algorithms expect numeric data to perform computations with.

Run the code cell below to separate the student data into feature and target columns to see if any features are non-numeric.

In [5]:
# Extract feature columns
feature_cols = list(student_data.columns[:-1])

# Extract target column 'passed'
target_col = student_data.columns[-1]

# Show the list of columns
print "Feature columns:\n{}".format(feature_cols)
print "\nTarget column: {}".format(target_col)

# Separate the data into feature data and target data (X_all and y_all, respectively)
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# Show the feature information by printing the first five rows
print "\nFeature values:"
print X_all.head()

Feature columns:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Target column: passed

Feature values:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...       

### Preprocess Feature Columns

As you can see, there are several non-numeric columns that need to be converted! Many of them are simply `yes`/`no`, e.g. `internet`. These can be reasonably converted into `1`/`0` (binary) values.

Other columns, like `Mjob` and `Fjob`, have more than two values, and are known as _categorical variables_. The recommended way to handle such a column is to create as many columns as possible values (e.g. `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc.), and assign a `1` to one of them and `0` to all others.

These generated columns are sometimes called _dummy variables_, and we will use the [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.

In [6]:
def preprocess_features(X):
    ''' Preprocesses the student data and converts non-numeric binary variables into
        binary (0/1) variables. Converts categorical variables into dummy variables. '''
    
    # Initialize new output DataFrame
    output = pd.DataFrame(index = X.index)

    # Investigate each feature column for the data
    for col, col_data in X.iteritems():
        
        # If data type is non-numeric, replace all yes/no values with 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # If data type is categorical, convert to dummy variables
        if col_data.dtype == object:
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Collect the revised columns
        output = output.join(col_data)
    
    return output

X_all = preprocess_features(X_all)
print "Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns))

Processed feature columns (48 total features):
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


### Implementation: Training and Testing Data Split
So far, we have converted all _categorical_ features into numeric values. For the next step, we split the data (both features and corresponding labels) into training and test sets. In the following code cell below, you will need to implement the following:
- Randomly shuffle and split the data (`X_all`, `y_all`) into training and testing subsets.
  - Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).
  - Set a `random_state` for the function(s) you use, if provided.
  - Store the results in `X_train`, `X_test`, `y_train`, and `y_test`.

In [10]:
# TODO: Import any additional functionality you may need here
from sklearn.cross_validation import train_test_split
# TODO: Set the number of training points
num_train = 300

# Set the number of testing points
num_test = X_all.shape[0] - num_train

# TODO: Shuffle and split the dataset into the number of training and testing points above
X_train, X_test, y_train, y_test = train_test_split(X_all, y_all, test_size=num_test, train_size=num_train, random_state=1)

# Show the results of the split
print "Training set has {} samples.".format(X_train.shape[0])
print "Testing set has {} samples.".format(X_test.shape[0])

Training set has 300 samples.
Testing set has 95 samples.


## Training and Evaluating Models
In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in `scikit-learn`. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to varying sizes of training data (100 data points, 200 data points, and 300 data points) and measure the F<sub>1</sub> score. You will need to produce three tables (one for each model) that shows the training set size, training time, prediction time, F<sub>1</sub> score on the training set, and F<sub>1</sub> score on the testing set.

**The following supervised learning models are currently available in** [`scikit-learn`](http://scikit-learn.org/stable/supervised_learning.html) **that you may choose from:**
- Gaussian Naive Bayes (GaussianNB)
- Decision Trees
- Ensemble Methods (Bagging, AdaBoost, Random Forest, Gradient Boosting)
- K-Nearest Neighbors (KNeighbors)
- Stochastic Gradient Descent (SGDC)
- Support Vector Machines (SVM)
- Logistic Regression

### Question 2 - Model Application
*List three supervised learning models that are appropriate for this problem. For each model chosen*
- Describe one real-world application in industry where the model can be applied. *(You may need to do a small bit of research for this — give references!)* 
- What are the strengths of the model; when does it perform well? 
- What are the weaknesses of the model; when does it perform poorly?
- What makes this model a good candidate for the problem, given what you know about the data?

**Answer: **

### Support Vector Machines

1. #### Real-world application:

    An application of the SVM learning algorithm is **cancer diagnosis**. This is a classification task, where the features are gene information and the response variable is whether the organism has cancer. [Ref. D]

2. #### Strengths:
    1. SVM models are still effective when there are more features than observations. [Ref. A]
    2. SVM models are memory efficient (the model can be stored as a set of support vectors). [Ref. A]
    3. SVMs can be used for both regression and classification. [Ref. A]
    4. SVMs are versatile because of their kernel hyperparameter. [Ref. A]

3. #### Weaknesses:
    1. SVMs are sensitive to noise. Mislabeled observations tend to hurt the performance of SVM models more than other models. [Ref. 3]

4. #### Justification for this problem:
    1. When compared to the iris data set which has four features for 150 observations, this data set seems to have enough features (30) compared to its number of observations (395) to justify using SVM.
    2. The scikit-learn developers made a flowchart [Ref. B] that suggests using SVM for this dataset. The logic is as follows:
        1. More than 50 observations
        2. Predicting a category
        3. Possess labeled data
        4. Less than 100,000 observations
        5. Not text data
    3. This dataset is likely correctly labeled, which avoids one of the weaknesses of SVMs. This dataset is likely correctly labeled because if it were mislabeled, a student would complain and would have the mislabel corrected.

5. #### References:
    1. http://scikit-learn.org/stable/modules/svm.html
    
    2. http://scikit-learn.org/stable/tutorial/machine_learning_map/
    
    3. http://www-labs.iro.umontreal.ca/~pift6080/H09/documents/papers/svm_tutorial.ppt
    
    4. http://www3.ntu.edu.sg/home/elpwang/pdf_web/05_svm_basic.pdf

### K-Nearest Neighbors

1. #### Real-world application:

    K-Nearest Neighbor models have been used to classify **handwritten digits**. [Ref. A]

2. #### Strengths:
    1. KNN models are useful when the decision boundary is irregular (i.e. when the decision boundary is hard to fit with a parametrized model). [Ref. A]
    2. KNN models are lazy. This reduces training time.
    3. The KNN algorithm is easy to understand and easy to implement. [Ref. B]

3. #### Weaknesses:
    1. KNN models take up more memory. Instead of storing parameters, they store the entire dataset.
    2. KNN models are lazy. This increases query time.

4. #### Justification for this problem:
    1. Like with my choice of SVMs, I used the scikit-learn flowchart. [Ref. C]
    2. The decision boundary of this classification task might be irregular.

5. #### References:
    1. http://scikit-learn.org/stable/modules/neighbors.html
    2. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2243774/?page=2
    3. http://scikit-learn.org/stable/tutorial/machine_learning_map/

### Boosting

1. #### Real-world application:

    Boosting, a type of ensemble method, is used for **object categorization** (a subproblem of computer vision). [Ref. B]

2. #### Strengths:
    1. By using many classifiers at once, ensemble methods tend to reduce overfitting. [Ref. A]

3. #### Weaknesses:
    1. Boosting is highly sensitive to noise. [Ref. D]
        2. "[If] any non-zero fraction of the training data is mis-labeled, the boosting algorithm tries extremely hard to correctly classify these training examples, and fails to produce a model with accuracy better than 1/2." [Ref. C]
    2. Ensemble methods can take a long time to use, since they require the training of multiple models.
    3. Boosting requires enough data for each classifier to satisfy the weak learner condition. [Ref. E]

4. #### Justification for this problem:
    1. Ensemble methods help prevent overfitting at the cost of taking a long time to train. However, since 395 observations is a small dataset, the increased cost of training is likely small.
    2. Since the data is unlikely to be unlabeled, boosting's weakness to noise is less likely to matter.

5. #### References:
    1. http://scikit-learn.org/stable/modules/ensemble.html#bagging-meta-estimator
    2. https://en.wikipedia.org/wiki/Boosting_(machine_learning)#Boosting_for_multi-class_categorization
    3. https://en.wikipedia.org/wiki/Boosting_(machine_learning)#Criticism
    4. http://www.cs.columbia.edu/~rocco/Public/mlj9.pdf
    5. https://storage.googleapis.com/supplemental_media/udacityu/5446988865/Intro%20to%20Boosting.pdf

### Setup
Run the code cell below to initialize three helper functions which you can use for training and testing the three supervised learning models you've chosen above. The functions are as follows:
- `train_classifier` - takes as input a classifier and training data and fits the classifier to the data.
- `predict_labels` - takes as input a fit classifier, features, and a target labeling and makes predictions using the F<sub>1</sub> score.
- `train_predict` - takes as input a classifier, and the training and testing data, and performs `train_clasifier` and `predict_labels`.
 - This function will report the F<sub>1</sub> score for both the training and testing data separately.

In [14]:
def train_classifier(clf, X_train, y_train):
    ''' Fits a classifier to the training data. '''
    
    # Start the clock, train the classifier, then stop the clock
    start = time()
    clf.fit(X_train, y_train)
    end = time()
    
    # Print the results
    print "Trained model in {:.4f} seconds".format(end - start)

    
def predict_labels(clf, features, target):
    ''' Makes predictions using a fit classifier based on F1 score. '''
    
    # Start the clock, make predictions, then stop the clock
    start = time()
    y_pred = clf.predict(features)
    end = time()
    
    # Print and return results
    print "Made predictions in {:.4f} seconds.".format(end - start)
    return f1_score(target.values, y_pred, pos_label='yes')


def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' Train and predict using a classifer based on F1 score. '''
    
    # Indicate the classifier and the training set size
    print "Training a {} using a training set size of {}. . .".format(clf.__class__.__name__, len(X_train))
    
    # Train the classifier
    train_classifier(clf, X_train, y_train)
    
    # Print the results of prediction for both training and testing
    print "F1 score for training set: {:.4f}.".format(predict_labels(clf, X_train, y_train))
    print "F1 score for test set: {:.4f}.".format(predict_labels(clf, X_test, y_test))
    print

### Implementation: Model Performance Metrics
With the predefined functions above, you will now import the three supervised learning models of your choice and run the `train_predict` function for each one. Remember that you will need to train and predict on each classifier for three different training set sizes: 100, 200, and 300. Hence, you should expect to have 9 different outputs below — 3 for each model using the varying training set sizes. In the following code cell, you will need to implement the following:
- Import the three supervised learning models you've discussed in the previous section.
- Initialize the three models and store them in `clf_A`, `clf_B`, and `clf_C`.
 - Use a `random_state` for each model you use, if provided.
 - **Note:** Use the default settings for each model — you will tune one specific model in a later section.
- Create the different training set sizes to be used to train each model.
 - *Do not reshuffle and resplit the data! The new training points should be drawn from `X_train` and `y_train`.*
- Fit each model with each training set size and make predictions on the test set (9 in total).  
**Note:** Three tables are provided after the following code cell which can be used to store your results.

In [15]:
# TODO: Import the three supervised learning models from sklearn
# from sklearn import model_A
from sklearn.svm import SVC
# from sklearn import model_B
from sklearn.neighbors import KNeighborsClassifier
# from skearln import model_C
from sklearn.ensemble import AdaBoostClassifier

# TODO: Initialize the three models
clf_A = SVC(random_state=1)
clf_B = KNeighborsClassifier()
clf_C = AdaBoostClassifier(random_state=1)

# TODO: Execute the 'train_predict' function for each classifier and each training set size
# train_predict(clf, X_train, y_train, X_test, y_test)
classifiers = [clf_A, clf_B, clf_C]
training_sizes = range(100, 400, 100)
for clf in classifiers:
    for size in training_sizes:
        train_predict(clf, X_train[:size], y_train[:size], X_test[:size], y_test[:size])

Training a SVC using a training set size of 100. . .
Trained model in 0.0017 seconds
Made predictions in 0.0009 seconds.
F1 score for training set: 0.8591.
Made predictions in 0.0007 seconds.
F1 score for test set: 0.8333.

Training a SVC using a training set size of 200. . .
Trained model in 0.0034 seconds
Made predictions in 0.0023 seconds.
F1 score for training set: 0.8581.
Made predictions in 0.0011 seconds.
F1 score for test set: 0.8408.

Training a SVC using a training set size of 300. . .
Trained model in 0.0064 seconds
Made predictions in 0.0049 seconds.
F1 score for training set: 0.8584.
Made predictions in 0.0016 seconds.
F1 score for test set: 0.8462.

Training a KNeighborsClassifier using a training set size of 100. . .
Trained model in 0.0004 seconds
Made predictions in 0.0010 seconds.
F1 score for training set: 0.7883.
Made predictions in 0.0012 seconds.
F1 score for test set: 0.7727.

Training a KNeighborsClassifier using a training set size of 200. . .
Trained model in 

### Tabular Results
Edit the cell below to see how a table can be designed in [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables). You can record your results from above in the tables provided.

** Classifer 1 - SVM**  

| Training Set Size | Training Time (s) | Prediction Time (test) (s) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |0.0017|0.0007|0.8591|0.8333|
| 200               |0.0034|0.0011|0.8581|0.8408|
| 300               |0.0064|0.0016|0.8584|0.8462|

** Classifer 2 - KNN**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |0.0004|0.0012|0.7883|0.7727|
| 200               |0.0007|0.0015|0.8345|0.7971|
| 300               |0.0007|0.0018|0.8558|0.7681|

** Classifer 3 - AdaBoost**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |0.0599|0.0033|0.9624|0.6949|
| 200               |0.0650|0.0033|0.8633|0.7647|
| 300               |0.0709|0.0033|0.8578|0.8116|

## Choosing the Best Model
In this final section, you will choose from the three supervised learning models the *best* model to use on the student data. You will then perform a grid search optimization for the model over the entire training set (`X_train` and `y_train`) by tuning at least one parameter to improve upon the untuned model's F<sub>1</sub> score. 

### Question 3 - Choosing the Best Model
*Based on the experiments you performed earlier, in one to two paragraphs, explain to the board of supervisors what single model you chose as the best model. Which model is generally the most appropriate based on the available data, limited resources, cost, and performance?*

**Answer: **

SVM is the best model for this task, under the following assumptions:
1. The possible models are SVM, KNN, and AdaBoost.
2. The performance metrics defining "best" are computation time and predictive power (measured by F1 score).
3. The hyperparameters of each model are set to their default values, which are defined by the scikit-learn development team.

SVM is the best model for this task for the following reasons:
1. SVM has the **highest F1 testing score** for each training set size.
2. SVM has the **fastest prediction time** for each training set size.
3. SVM trains faster than AdaBoost with respect to each training set size.
4. The SVM's faster prediction times and drastically higher F1 testing scores justify choosing it over KNN, despite KNN's drastically faster training times.
    1. The models will only need to be **trained once per year**, while the models will need to be **used multiple times per year**. This makes training time less important than prediction time. Even without this comparison, the SVM's drastically higher F1 scores are enough to justify choosing it over KNN.

### Question 4 - Model in Layman's Terms
*In one to two paragraphs, explain to the board of directors in layman's terms how the final model chosen is supposed to work. Be sure that you are describing the major qualities of the model, such as how the model is trained and how the model makes a prediction. Avoid using advanced mathematical or technical jargon, such as describing equations or discussing the algorithm implementation.*

**Answer: **

Each student has a number of attributes; including their age, sex, school, and more. From these attributes, we'd like to predict whether a given student will pass their high school's final exam. Looking at students who have taken their exam, we will study the relationship between their attributes and whether they passed their exam. For example, we might find that students with internet access are more likely to pass their exam; or, we might find the opposite. Once we understand the relationship between a student's attributes and whether they will pass their exam, we can use that understanding to predict whether a student who hasn't taken their exam will pass; and we can use that prediction to guide are student intervention efforts.

### Implementation: Model Tuning
Fine tune the chosen model. Use grid search (`GridSearchCV`) with at least one important parameter tuned with at least 3 different values. You will need to use the entire training set for this. In the code cell below, you will need to implement the following:
- Import [`sklearn.grid_search.gridSearchCV`](http://scikit-learn.org/stable/modules/generated/sklearn.grid_search.GridSearchCV.html) and [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Create a dictionary of parameters you wish to tune for the chosen model.
 - Example: `parameters = {'parameter' : [list of values]}`.
- Initialize the classifier you've chosen and store it in `clf`.
- Create the F<sub>1</sub> scoring function using `make_scorer` and store it in `f1_scorer`.
 - Set the `pos_label` parameter to the correct value!
- Perform grid search on the classifier `clf` using `f1_scorer` as the scoring method, and store it in `grid_obj`.
- Fit the grid search object to the training data (`X_train`, `y_train`), and store it in `grid_obj`.

In [52]:
# TODO: Import 'GridSearchCV' and 'make_scorer'
from sklearn.grid_search import GridSearchCV
from sklearn.metrics import make_scorer

# TODO: Create the parameters list you wish to tune
parameters = {'kernel': ['linear', 'poly', 'rbf', 'sigmoid'], 'degree': range(1,6)}

# TODO: Initialize the classifier
clf = SVC(random_state=1)

# TODO: Make an f1 scoring function using 'make_scorer' 
f1_scorer = make_scorer(f1_score, pos_label='yes')

# TODO: Perform grid search on the classifier using the f1_scorer as the scoring method
grid_obj = GridSearchCV(clf, parameters, scoring=f1_scorer)

# TODO: Fit the grid search object to the training data and find the optimal parameters
grid_obj = grid_obj.fit(X_train, y_train)

# Get the estimator
clf = grid_obj.best_estimator_

# Report the final F1 score for training and testing after parameter tuning
print "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
print "Tuned model has a testing F1 score of {:.4f}.".format(predict_labels(clf, X_test, y_test))

Made predictions in 0.0049 seconds.
Tuned model has a training F1 score of 0.8584.
Made predictions in 0.0016 seconds.
Tuned model has a testing F1 score of 0.8462.


### Question 5 - Final F<sub>1</sub> Score
*What is the final model's F<sub>1</sub> score for training and testing? How does that score compare to the untuned model?*

**Answer: **

The final model's F1 scores were 0.8584 and 0.8462 for training and testing, respectively. Neither of the scores changed with respect to the untuned model.

Extra note:

I tried the KNN model with grid search with n_neighbors from 1 to 30 and the F1 testing score was 0.8684, which is higher than the tuned SVM's score.

> **Note**: Once you have completed all of the code implementations and successfully answered each question above, you may finalize your work by exporting the iPython Notebook as an HTML document. You can do this by using the menu above and navigating to  
**File -> Download as -> HTML (.html)**. Include the finished document along with this notebook as your submission.